packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,123 @@
+import { reaction } from "mobx"
+import { getParentToChildPath, resolvePath } from "../parent"
+import { applyPatches, onPatches, patchRecorder } from "../patch"
+import { isRootStore, registerRootStore, unregisterRootStore } from "../rootStore"
+import { clone } from "../snapshot"
+import { assertTweakedObject } from "../tweaker/core"
+import { failure } from "../utils"
+
+/**
+ * Callback function for `DraftManager.withDraft`.
+ */
+export type WithDraftCallback<T extends object> = (draftNode: T) => boolean | undefined | void
+
+/**
+ * Manager class returned by `draftMiddleware` that allows you to make changes to a draft copy of
+ * the original subtree and apply them to the original subtree or reject them.
+ */
+export class DraftManager {
+  /**
+   * The draft copy of the original subtree.
+   */
+  private readonly subtreeRootDraft: object
+
+  /**
+   * The internal `onPatches` disposer.
+   */
+  private readonly disposers: Array<() => void>
+
+  /**
+   * Creates an instance of `DraftManager`.
+   * Do not use directly, use `draftMiddleware` instead.
+   *
+   * @param subtreeRoot Subtree root target object.
+   */
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot, { generateNewIds: false })
+
+    let wasRS = false
+    const disposeReactionRS = reaction(
+      () => isRootStore(subtreeRoot),
+      isRS => {
+        if (isRS !== wasRS) {
+          wasRS = isRS
+          if (isRS) {
+            registerRootStore(this.subtreeRootDraft)
+          } else {
+            unregisterRootStore(this.subtreeRootDraft)
+          }
+        }
+      },
+      { fireImmediately: true }
+    )
+
+    const disposeOnPatches = onPatches(subtreeRoot, patches =>
+      applyPatches(this.subtreeRootDraft, patches)
+    )
+
+    this.disposers = [disposeReactionRS, disposeOnPatches]
+  }
+
+  /**
+   * Executes `fn` with a draft of `node`. The changes made to the draft in `fn` can be accepted,
+   * i.e. applied to the original subtree, or rejected.
+   *
+   * @typeparam T Object type.
+   * @param node Object for which to obtain a draft copy.
+   * @param fn Function that is called with a draft copy of `node`. Any changes made to the draft
+   * are applied to the original subtree when `fn` returns `true`. When `fn` returns `false`,
+   * `undefined` or nothing, the changes made to the draft are rejected.
+   */
+  public withDraft<T extends object>(node: T, fn: WithDraftCallback<T>): void {
```

Comment:
it just occurred to me that it might be worth to decorate this function with mobx action to ensure all done inside this function (e.g. applyPatches) (and inside the fn function) is a single transaction and is untracked

Reply:
If the `withDraft` method was decorated with `@action`, the result of a reaction triggered by a change in `fn` would not be available in `fn`, would it? Because the reaction would be executed after `fn` has returned. Then, I wouldn't be able use the result of a change made to the draft in order to determine whether the change should also be applied to the original tree or rejected.

I could run the loop around `applyPatches` in an adhoc action to at least apply (inverse) patches in a single transaction. As for `fn`, I think the user should remain in control what part of `fn` should be executed as an action. I'd add a note about this in the documentation that I'll write once we've settled on the implementation.

Reply:
Good, sounds like a good compromise :)

Reply:
Done in a70cdd3b448290798799445367d2e1251ca13509.

Reply:
Should I write a test that multiple patches are applied in a single transaction?

Reply:
No need I think, that would be testing mobx functionality, wouldn't it?

Reply:
Well, we know that `applyPatches` is now wrapped in `runInAction` and that this function is tested, but none of the current tests checks the behavior of applying patches in a single transaction. Do you know what I mean?

Reply:
What would the test be like, checking if a reaction triggers only once and only after the the multiple patches wrapped inside a single runAction are run?

If so, I see it no different from doing any other kind of mutation (mobx side test).

Maybe I'm missing something though... 😕

Reply:
> What would the test be like, checking if a reaction triggers only once and only after the the multiple patches wrapped inside a single runAction are run?

Yes, that would have been my suggestion.

> If so, I see it no different from doing any other kind of mutation (mobx side test).
>
> Maybe I'm missing something though... :confused:

Hm, I guess you're right, no test. :-)